### PR TITLE
Redesign selinux::module defined type

### LINKED ISSUE #146: selinux::module should not manage files in /usr
## Affected Puppet, Ruby, OS and module versions/distributions

- Module version:  0.5.1-rc0

## How to reproduce (e.g Puppet code you use)

add a `selinux::module` resource

## What are you seeing

the resource creates/manages a file in `/usr/share/selinux`. According FSH standard `/usr` is *Secondary hierarchy for read-only user data;*. 

In `/usr` (excluding `/usr/local`) I only expect files managed by the systems package manager.

https://en.wikipedia.org/wiki/Filesystem_Hierarchy_Standard#Directory_structure

## What behaviour did you expect instead

The puppet managed file should be placed somewhere in `/var`. Maybe in some folder in puppets `$vardir` or whatever best practice is.

## references

* Depends on #178 (Redesign selinux::module parameters)

### LINKED ISSUE #178: Redesign selinux::module parameters
Current situation
-----------------------

* it only allows to build modules with only *.te files. no *.if and .fc extensions are possible
* it does not allow to install an already compiled module
* with some policies it's not needed to install make and co. it may install too much software not required and not wishfull on some systems.

Proposed interface
--------------------------

```
# only minimal module building (checkmodule, semodule_package)
selinux::module { 'mymodule':
  builder => 'minimal',
  source_te => 'puppet:///modules/profile/mymodule.te',
} 

# this will install additional required packages
# and build with make -f /usr/share/selinux/devel/Makefile
selinux::module { 'mymodule':
  ensure => 'present',
  builder => 'makefile', # DEFAULT 
  source_te => 'puppet:///modules/profile/mymodule.te',
  source_if => 'puppet:///modules/profile/mymodule.if',
  source_fc => 'puppet:///modules/profile/mymodule.fc',
} 

# no compiling at all, only loading an already compiled binary
selinux::module { 'mymodule':
  source_pp => 'puppet:///modules/profile/mymodule.pp',
} 

# remove a module
selinux::module { 'mymodule':
  ensure => 'absent',
} 

selinux::module { 'mymodule':
  ensure => 'disabled',
} 
```

Things to consider
-------------------------

* newer semodule utilities have the ability to specify a module priority (`-X,--priority=PRIORITY`) which isn't supported by the selmodule type. semodule defaults to priority 400 if nothing is defined (seen on EL7, Fedora 25)
* semodule provides the ability to disable and enable modules which the selmodule type does not support, valid usecase was given in #60 
* Ensure puppet agent throws errors if build or load of modules fails . It must fail every run until build and loading of the module succeeds

### PR DESCRIPTION
**Pull request by vinzent**
Implements whats discussed in  #178. 

* Adds ability to specify a builder (simple or refpolicy). Simple does not require the make utility. 
* Stops building in /usr - uses puppet agents vardir/puppet-selinux/

Extends and closes #189 
Closes #178 
Closes #146

### CODE REVIEW

**Comment by oranenj** [comment]
seems fine to me. Does the single-dir-building still work idempotently and correctly remove the sources when puppet stops managing a resource?

File: manifests/module.pp [hunk 1]
```diff
@@ -44,11 +53,11 @@
   validate_absolute_path($::selinux::config::module_build_dir)
   validate_absolute_path($::selinux::refpolicy_makefile)
 
-  $module_dir = "${::selinux::config::module_build_dir}/${title}"
+  $module_dir = $::selinux::config::module_build_dir
```
Comment (oranenj) [inline manifests/module.pp]: this and the above probably should be `$::selinux::module_build_dir` if the below is as well

File: manifests/module.pp [hunk 2]
```diff
@@ -44,11 +53,11 @@
...
   $module_file = "${module_dir}/${title}"
```
Comment (oranenj) [inline manifests/module.pp]: `$::selinux::config::module_build_dir` should be `$::selinux::module_build_dir`?

Comment (oranenj) [inline manifests/module.pp]: There's another instance in validate_absolute_path that didn't get highlighted

Comment (vinzent) [inline manifests/module.pp]: `::selinux::config::module_build_dir` != `::selinux::module_build_root` and `::selinux::module_build_dir` does not exist.

you created this var here: https://github.com/voxpupuli/puppet-selinux/pull/189/files#diff-ca85ef38c9b99f20d2d1c8d646a10e77R47 - it confused me too know. maybe we could come up with something more intuitive?  ;-)

**Comment by vinzent** [comment]
> Does the single-dir-building still work idempotently and correctly remove the sources when puppet stops managing a resource?

I suspect a problem when you manage policy A with .te and .fc and then update the manifest to only manage the .te. Then the .fc file will not get removed.

File: manifests/module.pp (hunk 2 above)
Comment (oranenj) [inline manifests/module.pp]: ah, yeah, the naming is too close together... Hm. is module_source_dir any better?

**Comment by oranenj** [comment]
@vinzent, do you think there's still something this needs, or could this be merged? I think the interface is fine as it is at least for my usecases, and also allows for backwards-compatible extension later on if needed.

**Comment by vinzent** [comment]
> @vinzent <https://github.com/vinzent>, do you think there's still
> something this needs
I think it can be merged.

I've  squashed the commits to one for you and one of mine. so history
looks better. what do you think?

**Comment by oranenj** [comment]
@vinzent: looks good to me.

**Comment by vinzent** [comment]
added a small fix for the selinux_module_refpolicy acceptance test. Upcoming selinux tools 2.6 (in f26) will not allow to grant usr_t some permissions (which is a good check). 

Re-run the acceptance tests on CentOS6, CentOS7, Fedora 24 and Fedora 25. All green.

**Comment by oranenj** [comment]
I added a duplicate label to get this out of the changelog in favour of the issue

**Pull request merged**